keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,314 @@
+  uid: 052fb0f4-3d50-11e5-b066-42010af0d7b6
+secrets:
+- name: default-token-uudge
+imagePullSecrets:
+- name: myregistrykey
+imageDecryptSecrets:
+- name: <secret name>
+- image: [<image name>]
+```
+
+For example,
+```bash
+kubectl patch serviceaccount default -p '{"imageDecryptSecrets":[{"name":"keysecret", "images":["localhost:5000/nginx:enc"]}]}'
+```
+
+#### Image Handler
+
+The privake keys are extracted from `ImageDecryptSecret` and passed to the CRI as by bundling them in `DecryptParams` as a part of `PullImageRequest`:
+
+
+```protobuf
+// DecryptParams is used to send decryption parameters to the image service of the CRI
+message DecryptParams {
+    repeated string privateKeyPasswds = 1;
+}
+```
+```protobuf
+message PullImageRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Authentication configuration for pulling the image.
+    AuthConfig auth = 2;
+    // Config of the PodSandbox, which is used to pull image in PodSandbox context.
+    PodSandboxConfig sandbox_config = 3;
+    // DecryptParams for the images service of the CRI
+    DecryptParams dcparams = 4;
+}
+```
+
+See the [reference implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L1063)
+
+Just like in the case of `ImagePullSecrets`, CRI [forwards](https://github.com/stefanberger/containerd/blob/image-encryption.v4/vendor/github.com/containerd/cri/pkg/server/image_pull.go#L103) the private keys to containerd (or CRI-O) where the actual decryption of the image will take place. 
+
+
+### Consumption of the ImageDecryptSecrets
+
+In the diagram https://imgur.com/a/eVkUF9w, when the kubelet wants to create a new pod which has encrypted images it has to first retrieve the decryption keys referenced decryption keys (They can be referenced directly in the pod or deployment yaml or in the pod’s service account). 
+
+Kubelet sends the request to pull the image along with the decryption keys to CRI which then forwards it to containerd. Containerd looks up for the image in the snapshotter used by CRI, if the image doesn’t exist then it downloads it and decrypts it using the decryption keys that were passed by CRI. 
+
+In case containerd finds the image in the snapshotter cache which was already pulled and decrypted by earlier request, it performs a check we call ‘Image Authorization’. This check only attempts to unwrap the keys in the image manifest of the image. In simple terms it means everytime you want to use an encrypted image, you will have to prove that you have the necessary keys to decrypt it even if the actual image is present in the decrypted form in the snapshotter. This prevents an attack where a user, without having decryption keys, might get access to encrypted image content if their pod gets scheduled on a worker node where that particular encrypted image was already pulled and decrypted by earlier request. 
```

Comment:
i think this problem is similar to how image pull secrets can be abused today.

pod A schedules to node n1 and uses an image c1 that required pull secret
pod B schedules to node n1 and uses an image c1 but does not specify a pull secret

result: pod B is able to use image c1

the workaround for this has been to enforce image pull policy of always where that was a security concern.

i wonder if we could model a solution for this globally that covers both ability to decrypt and ability to pull as part of a common operation.

Reply:
>i wonder if we could model a solution for this globally that covers both ability to decrypt and ability to pull as part of a common operation.

Yea this is something we definitely want to solve for both secret types :). We decided on making the behavior of this similar to `ImagePullSecrets`, assuming that there was a use case for this behavior, and intermittently solve the issue with admission control like you pointed out.

We are thinking that once this goes in, we can submit a separate KEP that would try to address the Pull secrets and Decrypt secrets together since the mechanism to do so would probably be similar. :)

Reply:
@derekwaynecarr Thanks for your comments. I have updated the draft with a dedicated section for `imagePullPolicy` [here](https://github.com/kubernetes/enhancements/blob/cbeff9708e7763d478ab729be9826705c3555db6/keps/sig-node/20190517-image-decryption.md#Relationship-with-imagePullPolicy)

Reply:
@harche Can you elaborate more about how we can quickly verify whether a decryption key matches already pulled image in containerd? Does that need network?

If it is a quick local verification, we may want to add that into container creation, because that seems to be a different problem with image pull secrets.
Image pull secrets verifies whether a image can be pulled from registry, which is strongly associated with the `ImagePullPolicy`.
However, the `image decryption secrets` is only about the image itself, no matter in registry or cached locally, it may make sense to verify that when creating the container.

Reply:
@Random-Liu most definitely! The encryption specification in the image manifest is as follows. The layer blob is encrypted with a symmetric key `k_sym` to produce the encrypted layer. The key `k_sym`  is wrapped by a public key `K_pub` and put in the annotations of each layer in the manifest of the image. 

In order to identify if the provided secrets would be able to decrypt the image, we just attempt the decryption of the data in the manifest annotations. If this passes, authorization is considered granted. 

Do you think it should be a separate CRI call to `CheckAuthorization` or adding to the input of `CreateContainer`?

Also, if it helps, it looks something like this (image is an estimation of the actual spec):

![image](https://user-images.githubusercontent.com/3060102/61764041-c05c6300-ada5-11e9-9bf3-5dae15f4d1db.png)

Reply:
@Random-Liu The way a runtime (like containerd) verifies if the decryption key matches an already pulled image is, it reads the annotations that hold the wrapped key of the individual layers. If the containerd is able to unwrap the key found in the corresponding annotation then it concludes that the provided key is valid. This operation happens on the worker node and doesn't require a network connection.  

If we want to make sure that we always check for the authorization irrespective of the `imagePullPolicy` then we need to send private key down [CreateContainer](https://github.com/harche/kubernetes/blob/pr_branch/pkg/kubelet/kuberuntime/kuberuntime_container.go#L124) apart from existing [EnsureImageExists](https://github.com/harche/kubernetes/blob/pr_branch/pkg/kubelet/kuberuntime/kuberuntime_container.go#L95) call. 

The only reason the current implementation of `ImageDecryptSecrets` sends the private keys only through `EnsureImageExists` is to be consistent with the behavior of the `ImagePullSecrets`

But we can easily get rid of the constraints around `imagePullPolicy` by sending the private keys through `CreateContainer` as well to make sure no matter what the `imagePullPolicy` is, the authorization check will always be performed. Of course, we will need a [corresponding change](https://github.com/stefanberger/containerd/blob/image-encryption.v5/cmd/ctr/commands/run/run_unix.go#L172) in containerd too.

Reply:
@lumjjb @harche Thanks for the detailed explanation!

> Image pull secrets verifies whether a image can be pulled from registry, which is strongly associated with the ImagePullPolicy.
However, the image decryption secrets is only about the image itself, no matter it is in the registry or cached locally, it makes sense to verify that whenever using it. 

As I mentioned above, image decryption secrets are similar with image pull secrets, but still different. For image decryption secrets particularly, I think we should pass and verify the private keys in `CreateContainer`.

Reply:
Agreed - That sounds good - we will change it to include the passing of the keys in `CreateContainer` as well. :)